server/odc-service/src/main/java/com/oceanbase/odc/metadb/worksheet/CollaborationWorksheetRepository.java

```diff
@@ -0,0 +1,161 @@
+        return count(specs);
+    }
+
+    static Specification<CollaborationWorksheetEntity> getSpecs(Long projectId,
+            String path, Integer minLevelNumberFilter, Integer maxLevelNumberFilter, String nameLikeFilter) {
+        Specification<CollaborationWorksheetEntity> specs = Specification
+                .where(OdcJpaRepository.eq(CollaborationWorksheetEntity_.projectId, projectId))
+                .and(OdcJpaRepository.startsWith(CollaborationWorksheetEntity_.path, path));
+        boolean min = minLevelNumberFilter != null && minLevelNumberFilter > 0;
+        boolean max = maxLevelNumberFilter != null && maxLevelNumberFilter > 0;
+        if (min && max) {
+            if (minLevelNumberFilter.equals(maxLevelNumberFilter)) {
+                // This line of code does not use 'between' in order to use index:idx_project_id_level_num_path
+                // and improve query performance
+                specs = specs.and(OdcJpaRepository.eq(CollaborationWorksheetEntity_.pathLevel, maxLevelNumberFilter));
+            } else if (minLevelNumberFilter < maxLevelNumberFilter) {
+                specs = specs
+                        .and(OdcJpaRepository.between(CollaborationWorksheetEntity_.pathLevel, minLevelNumberFilter,
+                                maxLevelNumberFilter));
+            }
+        } else if (min) {
+            specs = specs.and(OdcJpaRepository.gte(CollaborationWorksheetEntity_.pathLevel, minLevelNumberFilter));
+        } else if (max) {
+            specs = specs.and(OdcJpaRepository.lte(CollaborationWorksheetEntity_.pathLevel, maxLevelNumberFilter));
+        }
+        if (nameLikeFilter != null && !nameLikeFilter.isEmpty()) {
+            specs = specs.and(OdcJpaRepository.like(CollaborationWorksheetEntity_.path, nameLikeFilter));
+        }
+        return specs;
+    }
+
+    @Modifying
+    @Transactional
+    @Query("DELETE CollaborationWorksheetEntity c WHERE"
+            + " c.projectId = :projectId AND c.path LIKE CONCAT(:path,'%')")
+    int batchDeleteByPathLike(@Param("projectId") Long projectId, @Param("path") String path);
+
+    /**
+     * only for test
+     * 
+     * @param projectId
+     * @return
+     */
+    @Transactional
+    int deleteByProjectId(@Param("projectId") Long projectId);
+
+    @Modifying
+    @Transactional
+    @Query("UPDATE CollaborationWorksheetEntity c SET "
+            + " c.objectId = :#{#entity.objectId},c.size = :#{#entity.size},c.version = c.version + 1"
+            + " WHERE c.id = :#{#entity.id} AND c.version = :#{#entity.version}")
+    int updateContentByIdAndVersion(@Param("entity") CollaborationWorksheetEntity entity);
+
+    /**
+     * batch update path,not update value of other columns,especially columns such as version and
+     * object_id.
+     *
+     * @param idToPathMap id to new path map
+     * @return update rows count
+     */
+    @Transactional
+    default int batchUpdatePath(Map<Long, String> idToPathMap) {
```

Comment:
It is used in rename, up to 2000 data updates at once. 
If use batchUpdate, it will update objectId and version, It is possible to overwrite others' updates to the content of the worksheet.